Support setting Spring.webflux.base-path and supporting it in path predicate

Guofuyinan opened the pull request:
Fix Bug : Setting Spring.webflux.base-path All Path Predicates Return 404 Results

Issues : 
[New Issue : #2944](https://github.com/spring-cloud/spring-cloud-gateway/issues/2944)
[Three Years Ago Issue : #1759](https://github.com/spring-cloud/spring-cloud-gateway/issues/1759)

pivotal-cla commented:
@Guofuyinan Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-gateway&pullRequestId=2984)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-gateway&pullRequestId=2984) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-cla commented:
@Guofuyinan Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-gateway&pullRequestId=2984)!

spencergibb reviewed (changes requested):
please add a test

Guofuyinan commented:
> please add a test
I will add a test ASAP

spencergibb commented on spring-cloud-gateway-server/src/test/java/org/springframework/cloud/gateway/handler/predicate/GatewayPredicateVisitorTests.java:
```diff
@@ -16,25 +16,40 @@
 
 package org.springframework.cloud.gateway.handler.predicate;
 
-import java.util.ArrayList;
```
Changing the order of imports will cause a checkstyle failure

Guofuyinan replied:
This is my IDEA's automatic code formatting, and I will turn it off

spencergibb commented on spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/handler/predicate/PathRoutePredicateFactory.java:
```diff
@@ -114,8 +126,7 @@ public boolean test(ServerWebExchange exchange) {
 					return true;
-				}
-				else {
+				} else {
```
revert

Guofuyinan commented:
> Looks like some checkstyle issues.

Try again

Guofuyinan commented:
> Looks like some checkstyle issues.

Pls try again @spencergibb

Guofuyinan commented:
> Looks like some checkstyle issues.

Try again

scott-avery commented:
@spencergibb  any update?

spencergibb commented:
We will get this included in the 2024.0 release train.

Wangluyao99 commented:
> We will get this included in the 2024.0 release train.

@spencergibb  any update?

spencergibb commented:
Fixes #1759

Guofuyinan commented:
Extract variables

Guofuyinan commented:
@ryanjbaxter Any progress or anything I can do?

ryanjbaxter commented on spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/handler/predicate/PathRoutePredicateFactory.java:
```diff
@@ -82,7 +88,15 @@ public Predicate<ServerWebExchange> apply(Config config) {
+						basePath += ("/");
+					}
+				}
+				String pathPatternStr = basePathIsNotBlank ? basePath + pattern : pattern;
```
why not place this in previous `if` statement where we know `basePathIsNotBlank == true`

ryanjbaxter commented on spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/handler/predicate/PathRoutePredicateFactory.java:
```diff
@@ -50,8 +53,11 @@ public class PathRoutePredicateFactory extends AbstractRoutePredicateFactory<Pat
-	public PathRoutePredicateFactory() {
+	private final WebFluxProperties webFluxProperties;
+
+	public PathRoutePredicateFactory(WebFluxProperties webFluxProperties) {
```
Leave the previous constructor in place and mark it deprecated

ryanjbaxter commented:
Can you submit this against the 4.2.x branch?

Guofuyinan commented:
> Can you submit this against the 4.2.x branch?

done

Guofuyinan commented:
> Can you submit this against the 4.2.x branch?

i have changed the base to 4.2.x branch

ryanjbaxter commented:
One final thing, could you add a note to the documentation that this setting will impact the path predicate?

Guofuyinan commented:
> One final thing, could you add a note to the documentation that this setting will impact the path predicate?

Sorry, I found that there is no webflux related content under the docs of spring-cloud:4.2.x branch, while the main branch has it. Should I cherry-pick from the spring-cloud:main branch and modify it? 

If so, I plan to add the note to the 'The Path Route Predicate Factory' section in request-predicates-factories.adoc file.

Guofuyinan commented:
> I was thinking specifically here https://github.com/spring-cloud/spring-cloud-gateway/blob/4.2.x/docs/modules/ROOT/pages/spring-cloud-gateway/request-predicates-factories.adoc#the-path-route-predicate-factory.
> 
> It would be good to have this documented before we merge it so everything is in one PR

I've updated the documentation as suggested at the specific location you mentioned (`request-predicates-factories.adoc#the-path-route-predicate-factory`).
﻿
Added a paragraph explaining how the `spring.webflux.base-path` property influences path matching. The documentation now clearly states that this property value will be automatically prepended to path patterns, with an example included.
